src/coreclr/jit/assertionprop.cpp

```diff
@@ -3033,6 +3033,25 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* tree)
         }
         break;
 
+#if FEATURE_HW_INTRINSICS
+        case TYP_SIMD8:
+        case TYP_SIMD12:
+        case TYP_SIMD16:
+        case TYP_SIMD32:
+        {
+            assert(vnStore->IsVNVectorZero(vnCns));
+
+            VNFuncSimdTypeInfo vnInfo = vnStore->GetFuncSimdTypeOfVN(vnCns);
```

Comment:
Can you expand on this? I must be missing something because in order to create a SIMD zero node, you have to give it a base JIT type.

Reply:
> in order to create a SIMD zero node, you have to give it a base JIT type.

Right, and there is no way to get that for a `get_Zero` on XARCH. The fact it is possible on ARM64 is a side-effect of how conservatively is base type encoding handled there.

So what I am saying is that we should consider not decoding the base type at all here and instead use some fixed one, as the instruction to create the zero is the same for all base types (in essence, the base type does not matter).

It will enable this optimization for XARCH + make it possible to not encode base types for zeroes on ARM64 in the future without regressions. It should also simplify the change significantly.

(It is possible that in some future ISA the instruction to create `Vector<int>.Zero` would not be the same as for `Vector<float>.Zero`, so we should also guard against this possibility: `assert(arity == 0 || TARGET_ARM64)`)

Reply:
> for all base types (in essence, the base type does not matter).

I agree, and @tannergooding would agree with this as well. It would be nice to simply have a zero node where we do not care about the base type. At the moment, these nodes need to have a base type. If there was a dummy type, that would ideal. Nonetheless, we are not there yet; perhaps we could make a separate PR that makes that kind of change across the board. This PR is trying to be conservative.

> It should also simplify the change significantly.

I think it would simplify not only this change, but others as well, if we always assume SIMD zero node's base type doesn't matter. But again, I don't want to change too much in this PR. We could do this in a separate PR.

> Right, and there is no way to get that for a get_Zero on XARCH.

What does `GetSimdBaseJitType` return then?

Reply:
> I think it would simplify not only this change, but others as well, if we always assume SIMD zero node's base type doesn't matter. But again, I don't want to change too much in this PR. We could do this in a separate PR.

I see, that's fair. In that case I would suggest leaving comments around the new functions that they are workarounds for this issue (so that it does not get lost).

> What does `GetSimdBaseJitType` return then?

For nodes, it does of course return some type (used in the source). I meant that the VN for `get_Zero` functions on XARCH does not incorporate the base type.

Reply:
> In that case I would suggest leaving comments around the new functions that they are workarounds for this issue

I don't know if they are 'workarounds' necessarily but having a comment talking about this would be helpful for sure.

> For nodes, it does of course return some type (used in the source). I meant that the VN for get_Zero functions on XARCH does not incorporate the base type.

Based on that, this means that `encodesResultType` would be returning 'false' for the `get_Zero` nodes on XARCH where as on ARM64 it returns 'true'. Well, I could force it to encode the data if it's a vector zero. Also, be aware that `IsVectorZero` is only going to return `true` for HW intrinsic get_Zero nodes and nothing else.

Reply:
> I meant that the VN for get_Zero functions on XARCH does not incorporate the base type

This is a side-effect of the logic in `vnEncodesResultTypeForHWIntrinsic` (https://github.com/dotnet/runtime/blob/78143965703ed08a634c7c9a70994debf8ead4e9/src/coreclr/jit/hwintrinsic.cpp#L211-L258) that basically says "if all the ins entries are the same, assume all types are the same".

This ends up "working" because `get_Zero` sets all types to `INS_xorps`.

Notably the logic is also incorrect because its not accounting for cases where all entries are `INS_invalid` and therefore their may be differences that are introduced in lowering for nodes that are special-cased.

Reply:
> Also, be aware that `IsVectorZero` is only going to return true for HW intrinsic `get_Zero` nodes and nothing else.

And only on ARM64 (which is correct) :) - that's why it feels like a workaround to me. But as we've decided it is fine to be that way for now.

> Notably the logic is also incorrect because its not accounting for cases where all entries are INS_invalid and therefore their may be differences that are introduced in lowering for nodes that are special-cased.

I was wondering what was going on with those `INS_invalid`... We may have some issues to investigate in this area.

It seems like it would be more robust (and cheaper to query) if we assumed all intrinsics needed the base type and then manually marked those which we know don't (such as `get_Zero`) with a flag.

Reply:
> It seems like it would be more robust (and cheaper to query) if we assumed all intrinsics needed the base type and then manually marked those which we know don't (such as get_Zero) with a flag.

Yes. This should likely be an explicit opt-in via the flags:
1. By default assume the base type is important
2. If it is marked with a flag, the base type can be dropped
3. Optionally we can special case or expose flags for other scenarios (one example is `Add` where sign doesn't matter, but size and integer vs fp does)

Reply:
> And only on ARM64 (which is correct) :)

Where are you getting this? `IsVectorZero` checks for `gtOper == GT_HWINTRINSIC`  and then handles `Vector128/Vector256` on xarch and `Vector64/128` on arm64.

Reply:
I mixed up `IsVectorZero` that is the `GenTree*` function with `IsVNVectorZero` that is the new VN function being added.

The latter does indeed only return `true` for the relevant `VNFunc`s on ARM64 because of the arity check.